src/shims/unix/linux/epoll.rs

```diff
@@ -166,25 +376,160 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
+        events_op: &OpTy<'tcx>,
         maxevents: &OpTy<'tcx>,
         timeout: &OpTy<'tcx>,
-    ) -> InterpResult<'tcx, Scalar> {
+        place: &'tcx MPlaceTy<'tcx>,
+    ) -> InterpResult<'tcx> {
         let this = self.eval_context_mut();
 
-        let epfd = this.read_scalar(epfd)?.to_i32()?;
-        let _events = this.read_scalar(events)?.to_pointer(this)?;
-        let _maxevents = this.read_scalar(maxevents)?.to_i32()?;
-        let _timeout = this.read_scalar(timeout)?.to_i32()?;
+        let epfd_value = this.read_scalar(epfd)?.to_i32()?;
+        let maxevents = this.read_scalar(maxevents)?.to_i32()?;
+        let event = this.deref_pointer_as(
+            events_op,
+            this.libc_array_ty_layout("epoll_event", maxevents.try_into().unwrap()),
+        )?;
+        let timeout = this.read_scalar(timeout)?.to_u32()?;
 
-        let Some(mut epfd) = this.machine.fds.get_mut(epfd) else {
-            return Ok(Scalar::from_i32(this.fd_not_found()?));
+        if epfd_value <= 0 {
+            let einval = this.eval_libc("EINVAL");
+            this.set_last_error(einval)?;
+            this.write_scalar(Scalar::from_i32(-1), place)?;
+            return Ok(());
+        }
+        let Some(epfd) = this.machine.fds.get_ref(epfd_value) else {
+            let result_value = this.fd_not_found()?;
+            this.write_scalar(Scalar::from_i32(result_value), place)?;
+            return Ok(());
+        };
+        let mut binding = epfd.borrow_mut();
+        let epoll_file_description = &mut binding
+            .downcast_mut::<Epoll>()
+            .ok_or_else(|| err_unsup_format!("non-epoll FD passed to `epoll_wait`"))?;
+
+        let binding = epoll_file_description.get_ready_list();
+        let ready_list = binding.borrow_mut();
+
+        if ready_list.is_empty() {
+            if timeout == 0 {
+                // Non-blocking with no notification returned.
+                this.write_scalar(Scalar::from_i32(0), place)?;
+                return Ok(());
+            } else {
+                // Blocking
+                let duration = Duration::from_secs(timeout.into());
+                this.block_thread(
+                    BlockReason::Epoll,
+                    Some((TimeoutClock::Monotonic, TimeoutAnchor::Relative, duration)),
+                    callback!(
+                        @capture<'tcx> {
+                        epfd_value: i32,
+                        place: &'tcx MPlaceTy<'tcx>,
+                        event: MPlaceTy<'tcx>,
+                        }
+                        @unblock = |this| {
+                                    this.blocking_epoll_callback(epfd_value, place, &event);
+                                    Ok(())
+                                }
+                        @timeout = |this| {
+                        this.write_scalar(Scalar::from_i32(0), place)?;
+                                    Ok(())
+                                }
+                    ),
```

Comment:
I know this is a WIP, but please do spend the effort of using proper indentation inside this macro. rustfmt can't format macros so this has to be done by hand; see the other `callback!` invocations for our usual style.

Reply:
Oh I didn't know that rustfmt can't format macros, nice to know that. I will pay attention to this.